### PR TITLE
build: buildkit section

### PR DESCRIPTION
**Pull request by crazy-max**
~needs https://github.com/moby/buildkit/pull/3220~
needs https://github.com/docker/cli/pull/3832

Adds "BuildKit" section describing the BuildKit project the most basic way with cross linking to existing pages. Also adds a "Custom Dockerfile syntax" page reworked from Dockerfile reference documentation.

Following these changes I also merged the "build enhancements", "enabling buildkit" and "creating a dockerfile" pages in build section with proper redirects and cross linking.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

### CODE REVIEW

File: build/extend/index.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+that allows defining the relationships between your build operations using Go
+language primitives. This gives you full power to run anything you can imagine,
+but will probably not be how most people will define their builds. Instead, we
+expect most users to use a frontend component or LLB nested invocation to run
+some prepared combination of build steps.
+
+A frontend is a component that takes a human-readable
+build format and converts it to LLB so BuildKit can execute it. Frontends can
+be distributed as images, and the user can target a specific version of a
+frontend that is guaranteed to work for the features used by their definition.
```
Comment (dvdksn) [inline build/extend/index.md:61]: for later: I think we will want to lift the description of "what's a BuildKit frontend" to a more prominent/visible place. I was confused about why it was called frontends until I read these paragraphs, and now it totally makes sense. Let's leave it here for now though!

**Review by dvdksn — COMMENTED** [review]
This addition, and especially the LLB description, is so damn good I am gonna print it, frame it, and put it on a wall. I have some rephrasing suggestions, that's all.

File: build/extend/dockerfile-frontend.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+For example:
+
+```dockerfile
+# syntax=docker/dockerfile:1
+# syntax=docker.io/docker/dockerfile:1
+# syntax=example.com/user/repo:tag@sha256:abcdef...
+```
+
+This defines the location of the Dockerfile syntax that is used to build the
+Dockerfile. The BuildKit backend allows to seamlessly use external
```
Comment (jedevc) [inline build/extend/dockerfile-frontend.md:26]:
```suggestion
Dockerfile. The BuildKit backend allows seamlessly using external
```

File: build/extend/dockerfile-frontend.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+with the stable releases, and follow the same versioning with the `-labs` suffix,
+for example:
+
+- `docker/dockerfile:labs` - latest release on labs channel.
+- `docker/dockerfile:1-labs` - same as `dockerfile:1` in the stable channel,
+  with labs features enabled.
+- `docker/dockerfile:1.2-labs` -  same as `dockerfile:1.2` in the stable
+  channel, with labs features enabled.
+- `docker/dockerfile:1.2.1-labs` - immutable: never updated. Same as
+  `dockerfile:1.2.1` in the stable channel, with labs features enabled.
```
[the remark this answers is not in the document]
Comment (jedevc) [reply build/extend/dockerfile-frontend.md:84]: As another option, `experimental features`.

File: build/extend/index.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+- Prioritize your build cache for automatic pruning
+
+Apart from many new features, the main areas BuildKit improves on the current
+experience are performance, storage management, and extensibility. From the
+performance side, a significant update is a new fully concurrent build graph
+solver. It can run build steps in parallel when possible and optimize out
+commands that don't have an impact on the final result. We have also optimized
+the access to the local source files. By tracking only the updates made to these
+files between repeated build invocations, there is no need to wait for local
+files to be read or uploaded before the work can begin.
```
[the remark this answers is not in the document]
Comment (crazy-max) [reply build/extend/index.md:30]: I think we should avoid talking about the "legacy" builder but the overall build experience. "skip redundant" is not entirely true, some are really optimized like merge/diff operations.

File: Dockerfile [hunk 5]
```diff
@@ -61,7 +61,7 @@ RUN --mount=type=bind,from=generate,source=/out,target=_site <<EOF
     --disable-external \
     --internal-domains="docs.docker.com,docs-stage.docker.com,localhost:4000" \
     --file-ignore="/^./_site/engine/api/.*$/,./_site/registry/configuration/index.html" \
-    --url-ignore="/^/docker-hub/api/latest/.*$/,/^/engine/api/v.+/#.*$/,/^/glossary/.*$/" > /results 2>&1
+    --url-ignore="/^/docker-hub/api/latest/.*$/,/^/engine/api/v.+/#.*$/,/^/glossary/.*$/,/^/develop/develop-images/build_enhancements/#new-docker-build-secret-information$/" > /results 2>&1
```
Comment (crazy-max) [inline Dockerfile:64]: Temporarily ignore build_enhancements linking to internal hash new-docker-build-secret-information in htmlproofer while waiting for changes in Dockerfile reference on BuildKit repo:

```
#0 0.096 - ./_site/engine/reference/builder/index.html
#0 0.096   *  linking to internal hash #new-docker-build-secret-information that does not exist (line 2547)
#0 0.096      <a href="/develop/develop-images/build_enhancements/#new-docker-build-secret-information">[“build images with BuildKit”](https://github.com/develop/develop-images/build_enhancements/#new-docker-build-secret-information)</a>
```

Comment (crazy-max) [reply Dockerfile:64]: Opened https://github.com/moby/buildkit/pull/3220

**Review by jedevc — DISMISSED** [review]
LGTM once the corresponding buildkit PR is merged.

**Comment by crazy-max** [comment]
```
Ran on 1842 files!


- ./_site/engine/reference/commandline/build/index.html
  *  linking to internal hash #buildkit that does not exist (line 849)
     <a href="/engine/reference/builder/#buildkit">BuildKit backend</a>
  *  linking to internal hash #buildkit that does not exist (line 1114)
     <a href="/engine/reference/builder/#buildkit">enable BuildKit</a>
  *  linking to internal hash #buildkit that does not exist (line 1156)
     <a href="/engine/reference/builder/#buildkit">enable BuildKit</a>
```

Oh more links need to be changed

File: build/building/packaging.md [hunk 6]
```diff
@@ -29,6 +29,21 @@ multi-layer image builds based on your unique configurations. Dockerfiles can
 start simple and grow with your needs and support images that require complex
 instructions. For all the possible instructions, see the [Dockerfile reference](../../engine/reference/builder.md).
 
+The default filename to use for a Dockerfile is `Dockerfile` (without a file
+extension). Using the default name allows you to run the `docker build` command
```
Comment (dvdksn) [inline build/building/packaging.md:33]: Remove parenthesis

```suggestion
The default filename to use for a Dockerfile is `Dockerfile`, without a file
extension. Using the default name allows you to run the `docker build` command
```

File: build/building/packaging.md [hunk 7]
```diff
@@ -80,16 +95,23 @@ EXPOSE 8000
 CMD flask run --host 0.0.0.0 --port 8000
 ```
 
-We start by specifying the [syntax directive](../../engine/reference/builder.md#syntax).
-It pins the exact version of the Dockerfile syntax we're using:
+The first line to add to a Dockerfile is a [`# syntax` parser directive](../../engine/reference/builder.md#syntax).
+While _optional_, this directive instructs the Docker builder what syntax to use
```
Comment (dvdksn) [inline build/building/packaging.md:99]: Remove italics

```suggestion
While optional, this directive instructs the Docker builder what syntax to use
```

File: language/python/build-images.md [hunk 8]
```diff
@@ -8,9 +8,11 @@ description: Learn how to build your first Docker image by writing a Dockerfile
...
 ## Prerequisites
 
-Work through the orientation and setup in Get started [Part 1](../../get-started/index.md) to understand Docker concepts.
-
-{% include guides/enable-buildkit.md %}
+* Some awareness of basic Docker concepts. If unsure, work through the orientation
+  and setup in Get started [Part 1](../../get-started/index.md).
+* Basic understanding of the [Dockerfile format](../../build/building/packaging.md#dockerfile).
+* Ensure you have [enabled BuildKit](../../build/buildkit/index.md#getting-started)
+  on your machine.
```
Comment (dvdksn) [inline language/python/build-images.md:15]:
```suggestion
* You understand basic Docker concepts, and you have completed the orientation
  and setup in Get started [Part 1](../../get-started/index.md).
* You're familiar with the [Dockerfile format](../../build/building/packaging.md#dockerfile).
* You have [enabled BuildKit](../../build/buildkit/index.md#getting-started)
  on your machine.
```

File: language/nodejs/build-images.md [hunk 9]
```diff
@@ -10,9 +10,11 @@ redirect_from:
...
 ## Prerequisites
 
-Work through the orientation and setup in Get started [Part 1](../../get-started/index.md) to understand Docker concepts.
-
-{% include guides/enable-buildkit.md %}
+* Some awareness of basic Docker concepts. If unsure, work through the orientation
+  and setup in Get started [Part 1](../../get-started/index.md).
+* Basic understanding of the [Dockerfile format](../../build/building/packaging.md#dockerfile).
+* Ensure you have [enabled BuildKit](../../build/buildkit/index.md#getting-started)
+  on your machine.
```
Comment (dvdksn) [inline language/nodejs/build-images.md:17]:
```suggestion
* You understand basic Docker concepts, and you have completed the orientation
  and setup in Get started [Part 1](../../get-started/index.md).
* You're familiar with the [Dockerfile format](../../build/building/packaging.md#dockerfile).
* You have [enabled BuildKit](../../build/buildkit/index.md#getting-started)
  on your machine.
```

File: language/java/build-images.md [hunk 10]
```diff
@@ -8,9 +8,11 @@ description: Learn how to build your first Docker image by writing a Dockerfile
...
 ## Prerequisites
 
-Work through the orientation and setup in Get started [Part 1](../../get-started/index.md){:target="_blank" rel="noopener" class="_"} to understand Docker concepts. Refer to the following section for Java prerequisites.
-
-{% include guides/enable-buildkit.md %}
+* Some awareness of basic Docker concepts. If unsure, work through the orientation
+  and setup in Get started [Part 1](../../get-started/index.md).
+* Basic understanding of the [Dockerfile format](../../build/building/packaging.md#dockerfile).
+* Ensure you have [enabled BuildKit](../../build/buildkit/index.md#getting-started)
+  on your machine.
```
Comment (dvdksn) [inline language/java/build-images.md:15]:
```suggestion
* You understand basic Docker concepts, and you have completed the orientation
  and setup in Get started [Part 1](../../get-started/index.md).
* You're familiar with the [Dockerfile format](../../build/building/packaging.md#dockerfile).
* You have [enabled BuildKit](../../build/buildkit/index.md#getting-started)
  on your machine.
```

File: language/golang/build-images.md [hunk 11]
```diff
@@ -15,6 +15,9 @@ redirect_from:
   is a good starting point, so go (pun intended) check it out.
 * Some awareness of basic Docker concepts. If unsure, work through the orientation
   and setup in Get started [Part 1](../../get-started/index.md).
+* Basic understanding of the [Dockerfile format](../../build/building/packaging.md#dockerfile).
+* Ensure you have [enabled BuildKit](../../build/buildkit/index.md#getting-started)
+  on your machine.
```
Comment (dvdksn) [inline language/golang/build-images.md:20]:
```suggestion
* You understand basic Docker concepts, and you have completed the orientation
  and setup in Get started [Part 1](../../get-started/index.md).
* You're familiar with the [Dockerfile format](../../build/building/packaging.md#dockerfile).
* You have [enabled BuildKit](../../build/buildkit/index.md#getting-started)
  on your machine.
```

**Review by dvdksn — COMMENTED** [review]
LGTM; bunch of nits

**Comment by crazy-max** [comment]
Thanks for the review @dvdksn and @jedevc. Will wait for https://github.com/docker/cli/pull/3832 to be merged to fix the htmlproofer issue.

**Review by dvdksn — APPROVED** [review]

**Pull request merged**